ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2835,194 @@ public sealed class Bot : IAsyncDisposable {
+
+			if (linkElements.Count == 0) {
+				return new HashSet<uint>(0);
+			}
+
+			HashSet<uint> result = new HashSet<uint>(linkElements.Count);
+
+			foreach (string? badgeUri in linkElements.Select(htmlNode => htmlNode.GetAttribute("href"))) {
+				if (string.IsNullOrEmpty(badgeUri)) {
+					ArchiLogger.LogNullError(nameof(badgeUri));
+
+					continue;
+				}
+
+				// URIs to foil badges are the same as for normal badges except they end with "?border=1"
+				string? appIDText = badgeUri.Replace("?border=1", "").Split('/', StringSplitOptions.RemoveEmptyEntries).Last();
+
+				if (!uint.TryParse(appIDText, out uint appID) || (appID == 0)) {
+					ArchiLogger.LogNullError(nameof(appID));
+
+					continue;
+				}
+
+				result.Add(appID);
+			}
+
+			return result;
+		}
+
+		private async Task SendCompletedSets() {
+			ISet<uint> appIDs = await GetPossiblyCompletedBadgeAppIDs().ConfigureAwait(false);
+
+			if (appIDs.Count == 0) {
+				// Could not extract appIds for craftable badges - No need to load the inventory at this point
+				return;
+			}
+
+			HashSet<Steam.Asset> inventory;
+
+			try {
+				inventory = await ArchiWebHandler.GetInventoryAsync(SteamID)
+					.Where(item => item.Tradable)
+					.Where(item => (item.Type == Steam.Asset.EType.TradingCard) || (item.Type == Steam.Asset.EType.FoilTradingCard))
+					.Where(item => appIDs.Contains(item.RealAppID))
+					.ToHashSetAsync()
+					.ConfigureAwait(false);
+			} catch (HttpRequestException e) {
+				ArchiLogger.LogGenericWarningException(e);
+
+				return;
+			} catch (Exception e) {
+				ArchiLogger.LogGenericException(e);
+
+				return;
+			}
+
+			if (inventory.Count == 0) {
+				ArchiLogger.LogGenericWarning(string.Format(Strings.ErrorIsEmpty), nameof(inventory));
+
+				return;
+			}
+
+			Dictionary<(uint RealAppID, Steam.Asset.EType Type, Steam.Asset.ERarity Rarity), uint> inventorySets = Trading.GetInventorySets(inventory).ToDictionary(entry => entry.Key, entry => entry.Value.FirstOrDefault());
+			HashSet<Steam.Asset> result = (await Utilities.InParallel(inventorySets.Select(async set => GetItemsForFullBadge(inventory, set.Key.RealAppID, set.Key.Type, set.Key.Rarity, await ArchiWebHandler.GetCardCountForGame(set.Key.RealAppID).ConfigureAwait(false), set.Value))).ConfigureAwait(false)).SelectMany(set => set).ToHashSet();
+
+			if (result.Count > 0) {
+				await Actions.SendInventory(result).ConfigureAwait(false);
+			}
+		}
+
+		internal static IEnumerable<Steam.Asset> GetItemsForFullBadge(IReadOnlyCollection<Steam.Asset> inventory, uint appID, Steam.Asset.EType type, Steam.Asset.ERarity rarity, byte cardsPerSet, uint amountPerCard) {
+			if ((cardsPerSet == 0) || (amountPerCard == 0)) {
+				yield break;
+			}
+
+			Dictionary<ulong, List<Steam.Asset>> itemsPerClassID = inventory.Where(item => item.RealAppID == appID)
```

Comment:
This is very suboptimal, per each appID you're going to go through whole inventory over and over again looking for it, then again per type and per rarity. So many CPU cycles and memory wasted only because you didn't decide to do it in one loop - rewrite it, there is no reason in the world why it can't accept a set of `(uint appID, Steam.Asset.EType type, Steam.Asset.ERarity rarity, byte cardsPerSet, uint setsToExtract)`.

Reply:
First you don't want me to create new dictionary and now you want me to?

Reply:
Do you see a difference in iterating over inventory once and creating one dictionary vs iterating over inventory 700 times for 700 appIDs and creating 700 dictionaries?

Reply:
yes